src/ray/core_worker/core_worker.cc

```diff
@@ -294,7 +294,43 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
   }
 
   core_worker_client_pool_ =
-      std::make_shared<rpc::CoreWorkerClientPool>(*client_call_manager_);
+      std::make_shared<rpc::CoreWorkerClientPool>([&](const rpc::Address &addr) {
+        return std::make_shared<rpc::CoreWorkerClient>(
+            addr,
+            *client_call_manager_,
+            /*core_worker_unavailable_timeout_callback=*/[this, addr]() {
+              const NodeID node_id = NodeID::FromBinary(addr.raylet_id());
+              const WorkerID worker_id = WorkerID::FromBinary(addr.worker_id());
+              const rpc::GcsNodeInfo *node_info =
+                  gcs_client_->Nodes().Get(node_id, /*filter_dead_nodes=*/true);
+              if (node_info == nullptr) {
+                RAY_LOG(INFO).WithField(worker_id).WithField(node_id)
+                    << "Disconnect core worker client since its node is dead";
+                io_service_.post(
+                    [this, worker_id]() {
+                      core_worker_client_pool_->Disconnect(worker_id);
+                    },
+                    "CoreWorkerClientPool.Disconnect");
+
+                return;
+              }
+
+              auto raylet_client = std::make_shared<raylet::RayletClient>(
```

Comment:
I don't understand why we need shared pointer? Doesn't plain type `RayletClient` suffice?

Reply:
Currently we cannot create plain type of `NodeManagerWorkerClient` as it's constructor is private. Don't want to change this in this PR.

Reply:
I mean, 
```
RayletClient raylet_client(rpc::NodeManagerWorkerClient::make(...));
```

Reply:
Oh, I see. Good catch.